### PR TITLE
Zanderz/core 3018

### PR DESCRIPTION
**Pull request by zanderz**
@maxtaco @taruti 
This is the major Wix effort; it can successfully install driver + user mode apps, then uprade the apps silently. We're not shipping this yet but it's an important milestone.

### CODE REVIEW

**Comment by zanderz** [comment]
@oconnor663 @chrisnojima 
Thanks for volunteering

**Comment by keybase-ci-visdiff** [comment]
:mag_right: No visual changes found as a result of these commits.

**Comment by keybase-ci-visdiff** [comment]
:mag_right: No visual changes found as a result of these commits.

**Comment by oconnor663** [comment]
I'm afraid I don't have any useful comments on the code. We ran through a test run just now with mixed results:

- The initial install failed with confusing error messages (and pieces of Windows desktop UI I honestly didn't know existed), because I had an ancient version of Keybase already on my machine.
- After I uninstalled my ancient version and tried again, everything worked.
- Everything was super slow, probably because I only had 512MB of RAM on this VM.

Some things I'd like to test at some point:

- What happens if explorer.exe has a Keybase folder open at the exact moment an update happens? Does it gracefully handle the errors?
- Does the install succeed if I have a more recent existing version of Keybase that it needs to pave over?

File: packaging/windows/WIXInstallers/KeybaseApps/KeybaseApps.wxs [hunk 1]
```diff
@@ -0,0 +1,278 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
+	<Product Id="*" Name="Keybase Update" Language="1033" Version="$(env.KEYBASE_WINVER)" Manufacturer="Keybase, Inc." UpgradeCode="c118f7ec-9a1d-4ff1-91f5-15d208499d7b">
+		<Package InstallerVersion="500" Compressed="yes" InstallPrivileges="limited" InstallScope="perUser" />
+    <MediaTemplate EmbedCab="yes"/>
+    <Icon Id="ProductIcon" SourceFile="$(env.GOPATH)\src\github.com\keybase\client\packaging\windows\keybase.ico"/>
+    <Property Id="ARPPRODUCTICON" Value="ProductIcon"/>
+    <Property Id="ARPHELPLINK" Value="http://www.keybase.io"/>
```
Comment (chrisnojima) [inline packaging/windows/WIXInstallers/KeybaseApps/KeybaseApps.wxs]: https? and below

File: packaging/windows/WIXInstallers/KeybaseApps/KeybaseApps.wxs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+      <Dialog Id="UpdateDlg" Width="260" Height="95" Title="[ProductName] [Setup]" NoMinimize="yes">
+        <Control Id="No" Type="PushButton" X="132" Y="67" Width="56" Height="17" Default="yes" Cancel="yes" Text="[ButtonText_No]">
+          <Publish Event="EndDialog" Value="Exit">1</Publish>
+        </Control>
+        <Control Id="Yes" Type="PushButton" X="72" Y="67" Width="56" Height="17" Text="[ButtonText_Yes]">
+          <Publish Event="EndDialog" Value="Return">1</Publish>
+        </Control>
+        <Control Id="Text" Type="Text" X="43" Y="15" Width="209" Height="30">
+          <Text>This [ProductName] update contains important bugfixes and cool new features. Would you like to install it now?</Text>
```
Comment (chrisnojima) [inline packaging/windows/WIXInstallers/KeybaseApps/KeybaseApps.wxs]: we need to talk about how this gets updated / removed

File: packaging/windows/WIXInstallers/KeybaseApps/KeybaseApps.wxs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+      <Component Id="runquiet.exe" Guid="{1265F74B-0775-4522-A67D-25422D1984A3}">
+        <CreateFolder />
+        <RemoveFolder Id="RemoveMyAppRunner" On="uninstall" />
+        <RegistryKey Root="HKCU" Key="Software\Keybase\Keybase">
+          <RegistryValue Name="runquietexe" Value="1" KeyPath="yes" Type="integer" />
+        </RegistryKey>
+        <File Id="runquiet.exe" Source="$(env.GOPATH)\src\github.com\keybase\client\go\runquiet\runquiet.exe" Checksum="yes"/>
+      </Component>
+
+      <!--<Component Id="DokanInstallerFile" Guid="{D0E508C1-9010-4D68-A3D4-CDCEE8D07BEF}">
```
Comment (chrisnojima) [inline packaging/windows/WIXInstallers/KeybaseApps/KeybaseApps.wxs]: 🔥

File: packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                 Condition="NOT DokanUninstallString"        
+                 Variable="DokanUninstallString"
+                 Root="HKLM"
+                 Key="SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{65A3A964-3DC3-0100-0000-160511152538}"
+                 Value="UninstallString"
+                 Format="raw"
+                 Win64="yes"
+                 />
+    <!--
+    <bal:Condition Message="Please uninstall previous Keybase version and reboot before proceeding with this update: [WixBundleOriginalSource]">
```
Comment (chrisnojima) [inline packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs]: 🔥

File: packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+          InstallCommand="/quiet"
+          UninstallCommand="/quiet /uninstall"
+          DetectCondition="ExeDetectedVariable"
+          PerMachine="yes"
+          Permanent="no">
+      </ExePackage>
+      
+      <!--
+      <ExePackage
+        SourceFile="$(env.GOPATH)\bin\dokan-dev\dokan-v0.8.0\DokanInstall_0.8.0_redist.exe"
```
Comment (chrisnojima) [inline packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs]: 🔥

**Comment by zanderz** [comment]
@oconnor663 I added more detection for more older installers; should cover your case.
KBFS: it does not currently gracefully wait for you to finish your current operation; we'll have to revisit that.
When you say "more recent version", if you mean you already have a newer one than the installer you're running, it lets you know that and quits (or at least it's supposed to).

**Comment by oconnor663** [comment]
Nice. My main worry for restarting KBFS while programs are using it isn't so much that their writes will fail. (Hopefully we can trust clients to handle write errors in some way?) It's more that some program will have set its current directory to KBFS, so that it ends up in a weird broken state from then on. Though maybe directory switching isn't as common on Windows? Anyway, no idea what the semantics are there.

**Comment by zanderz** [comment]
Can I merge? Again, it won't affect current builds.

**Comment by oconnor663** [comment]
No objections from me. @chrisnojima?

**Comment by gabriel** [comment]
updater.exe changes look ok to me

**Comment by chrisnojima** [comment]
do it

**Pull request merged**